c7n/resources/s3.py

```diff
@@ -839,6 +839,65 @@ def get_std_format_args(self, bucket):
         }
 
 
+@S3.filter_registry.register('lock-configuration')
+class S3LockConfigurationFilter(ValueFilter):
+    """
+    Filter S3 buckets based on their object lock configurations
+
+    :example:
+
+    Get all buckets where lock configuration mode is COMPLIANCE
+
+        .. code-block:: yaml
+
+                policies:
+                  - name: lock-configuration-compliance
+                    resource: aws.s3
+                    filters:
+                      - type: lock-configuration
+                        key: Rule.DefaultRetention.Mode
+                        value: COMPLIANCE
+
+    """
+    schema = type_schema('lock-configuration', rinherit=ValueFilter.schema)
+    permissions = ('s3:GetBucketObjectLockConfiguration',)
+    annotate = True
+    annotation_key = 'ObjectLockConfiguration'
```

Comment:
generally want a c7n: prefix on annotation keys to distinguish from native attributes.